runtime/compiler/env/J9ClassEnv.cpp

```diff
@@ -515,6 +515,15 @@ static void addEntryForFieldImpl(TR_VMField *field, TR::TypeLayoutBuilder &tlb,
       ListIterator<TR_VMField> iter(fieldsInfo.getFields());
       for (TR_VMField *childField = iter.getFirst(); childField; childField = iter.getNext())
          {
+         IDATA offsetBaseForChild = field->offset + offsetBase;
+         if (J9CLASS_HAS_4BYTE_PREPADDING(fieldClass))
```

Comment:
Why isn't this accounted for in the offset itself?

Reply:
There are two kinds of layout, the standalone layout and the flattened layout. For example given : `inline class ValueLong { long l; }` the standalone layout is:

```
ValueLong {
    clazz 
    /* 4byte pre-padding */
    long l (offset = 4) 
}

//note `instanceFieldOffset` returns offset from header not from start of the object
```

We need to insert prepadding in order to statisfy alignment constraints for the long. 

However if it were flattened within another type,`inline class ValueLongRef {ValueLong l; }` the layout would be:

```
ValueLongRef {
    clazz 
    /* 4byte pre-padding */
    ValueLong { (offset = 4 in ValueLongRef) 
         //Value Long flattened layout, no pre-padding
         long l (offset = 0 in ValueLong) //`instanceFieldOffset` still returns 4, this is why we need to subtract prepadding
   }
}
```

`instanceFieldOffset` always returns the offsets based on the standalone layout. So when a type is flattened we need to know if the standalone layout has pre-padding since that will be removed when its flattened.